src/niscope/templates/session.py/fetch_waveform.py.mako

```diff
@@ -13,47 +13,48 @@
         import numpy
         import sys
 
-        # Set the fetch attributes
-        with _NoChannel(session=self):
-            self._fetch_relative_to = relative_to
-            self._fetch_offset = offset
-            self._fetch_record_number = record_number
-            self._fetch_num_records = -1 if num_records is None else num_records
-
-        num_samples = int(len(waveform) / self._actual_num_wfms())
-
-        if waveform.dtype == numpy.float64:
-            wfm_info = self._fetch_into_numpy(num_samples=num_samples, waveform=waveform, timeout=timeout)
-        elif waveform.dtype == numpy.int8:
-            wfm_info = self._fetch_binary8_into_numpy(num_samples=num_samples, waveform=waveform, timeout=timeout)
-        elif waveform.dtype == numpy.int16:
-            wfm_info = self._fetch_binary16_into_numpy(num_samples=num_samples, waveform=waveform, timeout=timeout)
-        elif waveform.dtype == numpy.int32:
-            wfm_info = self._fetch_binary32_into_numpy(num_samples=num_samples, waveform=waveform, timeout=timeout)
-        else:
-            raise TypeError("Unsupported dtype. Is {0}, expected {1}, {2}, {3}, or {5}".format(waveform.dtype, numpy.float64, numpy.int8, numpy.int16, numpy.int32))
-
-        if sys.version_info.major >= 3:
-            # In Python 3 and newer we can use memoryview objects to give us pieces of the underlying array. This is much faster
-            mv = memoryview(waveform)
-
-        i = 0
-        lwfm_i = len(wfm_info)
-        lrcl = len(self._repeated_capability_list)
-        assert lwfm_i % lrcl == 0, 'Number of waveforms should be evenly divisible by the number of channels: len(wfm_infos) == {0}, len(self._repeated_capability_list) == {1}'.format(lwfm_i, lrcl)
-        actual_num_records = int(lwfm_i / lrcl)
-        for chan in self._repeated_capability_list:
-            for rec in range(offset, offset + actual_num_records):
-                wfm_info[i].channel = chan
-                wfm_info[i].record = rec
-
-                if sys.version_info.major >= 3:
-                    start = i * num_samples
-                    end = start + wfm_info[i].actual_samples
-                    del wfm_info[i].actual_samples
-                    wfm_info[i].samples = mv[start:end]
-
-                i += 1
-
-        return wfm_info
+        with self.lock():
```

Comment:
While this is cleaner than looking at config, it divorces it from the metadata setting of whether to use the lock or not.
Hopefully this can be all fixed through the use of decorators instead so it becomes a non-issue.